<a href="https://colab.research.google.com/github/SallyLandon/sds510/blob/main/Module3basic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving crimestat (1).csv to crimestat (1).csv


In [ ]:
import pandas as pd

df = pd.read_csv('crimestat (1).csv', escapechar='\\', quoting=3) # quoting=3 means QUOTE_NONE, Gemini debug
# Remove quotes from column names
df.columns = df.columns.str.replace('"', '')
df.head()

,INC NUMBER,OCCURRED ON,OCCURRED TO,UCR CRIME CATEGORY,100 BLOCK ADDR,ZIP,PREMISE TYPE
0,"""201600000052855""",11/01/2015 00:00,01/09/2016 00:00,"""MOTOR VEHICLE THEFT""","""N 43RD AVE & W CACTUS RD""","""85029""","""SINGLE FAMILY HOUSE"""
1,"""201600000594484""",11/01/2015 00:00,NaN,"""RAPE""","""13XX E ALMERIA RD""","""85006""","""SINGLE FAMILY HOUSE"""
2,"""201500002102327""",11/01/2015 00:00,11/01/2015 09:00,"""LARCENY-THEFT""","""51XX N 15TH ST""","""85014""","""APARTMENT"""
3,"""201500002101405""",11/01/2015 00:00,11/01/2015 05:00,"""MOTOR VEHICLE THEFT""","""102XX W MEDLOCK AVE""","""85307""","""SINGLE FAMILY HOUSE"""
4,"""201500002102668""",11/01/2015 00:00,11/01/2015 11:50,"""MOTOR VEHICLE THEFT""","""69XX W WOOD ST""","""85043""","""SINGLE FAMILY HOUSE"""


I wanted to look at the structure of the dataset.  There are 7 columns as listed below:

In [ ]:
df.head(10)

,INC NUMBER,OCCURRED ON,OCCURRED TO,UCR CRIME CATEGORY,100 BLOCK ADDR,ZIP,PREMISE TYPE
0,"""201600000052855""",11/01/2015 00:00,01/09/2016 00:00,"""MOTOR VEHICLE THEFT""","""N 43RD AVE & W CACTUS RD""","""85029""","""SINGLE FAMILY HOUSE"""
1,"""201600000594484""",11/01/2015 00:00,NaN,"""RAPE""","""13XX E ALMERIA RD""","""85006""","""SINGLE FAMILY HOUSE"""
2,"""201500002102327""",11/01/2015 00:00,11/01/2015 09:00,"""LARCENY-THEFT""","""51XX N 15TH ST""","""85014""","""APARTMENT"""
3,"""201500002101405""",11/01/2015 00:00,11/01/2015 05:00,"""MOTOR VEHICLE THEFT""","""102XX W MEDLOCK AVE""","""85307""","""SINGLE FAMILY HOUSE"""
4,"""201500002102668""",11/01/2015 00:00,11/01/2015 11:50,"""MOTOR VEHICLE THEFT""","""69XX W WOOD ST""","""85043""","""SINGLE FAMILY HOUSE"""
5,"""201600000527709""",11/01/2015 00:00,03/22/2016 00:36,"""LARCENY-THEFT""","""33XX W CAMELBACK RD""","""85017""","""PARKING LOT"""
6,"""201700001722914""",11/01/2015 00:00,NaN,"""LARCENY-THEFT""","""279XX N 23RD LN""","""85085""","""SINGLE FAMILY HOUSE"""
7,"""201700001603695""",11/01/2015 00:00,03/31/2016 00:00,"""RAPE""","""38XX W CAMELBACK RD""","""85019""","""HOSPITAL"""
8,"""201500002168686""",11/01/2015 00:00,11/11/2015 09:30,"""LARCENY-THEFT""","""14XX E HIGHLAND AVE""","""85014""","""PARKING LOT"""
9,"""201500002103724""",11/01/2015 00:00,11/01/2015 15:21,"""DRUG OFFENSE""","""54XX W INDIAN SCHOOL RD""","""85031""","""APARTMENT"""


In [ ]:
df.tail(10)

,INC NUMBER,OCCURRED ON,OCCURRED TO,UCR CRIME CATEGORY,100 BLOCK ADDR,ZIP,PREMISE TYPE
131667,"""201700002024361""",11/17/2017 03:00,11/17/2017 03:20,"""AGGRAVATED ASSAULT""","""6XX E TOWNLEY AVE""","""85020""","""APARTMENT"""
131668,"""201780009020""",11/17/2017 03:13,11/17/2017 03:20,"""LARCENY-THEFT""","""206XX N 31ST DR""","""85027""","""SINGLE FAMILY HOUSING"""
131669,"""201700002024345""",11/17/2017 03:17,NaN,"""LARCENY-THEFT""","""59XX E SHEENA DR""","""85254""","""CARPORT"""
131670,"""201700002024346""",11/17/2017 03:19,11/17/2017 03:19,"""DRUG OFFENSE""","""41XX N 27TH AVE""","""85017""","""VEHICLE"""
131671,"""201700002024419""",11/17/2017 04:00,11/17/2017 04:06,"""BURGLARY""","""22XX E BROADWAY RD""","""85040""","""DEPARTMENT / DISCOUNT STORE"""
131672,"""201700002027239""",11/17/2017 04:15,11/17/2017 04:30,"""LARCENY-THEFT""","""16XX E INDIAN SCHOOL RD""","""85016""","""CONVENIENCE MARKET / STORE"""
131673,"""201700002025582""",11/17/2017 05:00,11/17/2017 09:50,"""MOTOR VEHICLE THEFT""","""43XX N 103RD AVE""","""85037""","""APARTMENT"""
131674,"""201700002024523""",11/17/2017 05:34,NaN,"""AGGRAVATED ASSAULT""","""8XX N 59TH AVE""","""85043""","""ABANDONED/CONDEMNED STRUCTURE"""
131675,"""201700002030009""",11/17/2017 05:50,11/17/2017 07:50,"""BURGLARY""","""51XX N 39TH DR""","""85019""","""SINGLE FAMILY HOUSE"""
131676,"""201700002107372""",11/17/2017 06:00,11/30/2017 06:00,"""BURGLARY""","""68XX N 25TH DR""","""85017""","""A"


In [ ]:
df.sample(10)

,INC NUMBER,OCCURRED ON,OCCURRED TO,UCR CRIME CATEGORY,100 BLOCK ADDR,ZIP,PREMISE TYPE
14116,"""201600000156294""",01/24/2016 08:00,01/25/2016 23:50,"""BURGLARY""","""8XX E PEORIA AVE""","""85020""","""SINGLE FAMILY HOUSE"""
74844,"""201700000041957""",NaN,01/08/2017 00:10,"""AGGRAVATED ASSAULT""","""33XX W MCLELLAN BLVD""","""85017""","""SINGLE FAMILY HOUSE"""
2952,"""201500002224163""",11/19/2015 10:00,11/19/2015 20:00,"""BURGLARY""","""20XX E MITCHELL DR""","""85016""","""SINGLE FAMILY HOUSE"""
110788,"""201700001372013""",07/25/2017 22:00,NaN,"""RAPE""","""82XX W MCDOWELL RD""","""85043""","""APARTMENT"""
110796,"""201700001298358""",07/25/2017 22:18,NaN,"""AGGRAVATED ASSAULT""","""67XX W INDIAN SCHOOL RD""","""85033""","""GAS / SERVICE STATION"""
17750,"""201600000283399""",02/14/2016 13:02,NaN,"""LARCENY-THEFT""","""172XX N 19TH AVE""","""85023""","""RETAIL BUSINESS"""
75179,"""201700000056645""",01/10/2017 00:00,NaN,"""DRUG OFFENSE""","""N 13TH AVE & W POLK ST""","""85007""","""STREET / ROADWAY / ALLEY / SIDEWALK"""
49025,"""201600001513084""",08/15/2016 17:35,NaN,"""DRUG OFFENSE""","""186XX N 19TH AVE""","""85027""","""RETAIL BUSINESS"""
96584,"""201700000811580""",05/10/2017 13:30,05/10/2017 13:30,"""LARCENY-THEFT""","""13XX E BELL RD""","""85022""","""GROCERY / SUPER MARKET"""
111044,"""201700001345829""",07/27/2017 11:00,08/02/2017 15:00,"""BURGLARY""","""20XX N 17TH ST""","""85006""","""SINGLE FAMILY HOUSE"""


In [ ]:
df.info()
df.sample(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131677 entries, 0 to 131676
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   INC NUMBER          131677 non-null  object
 1   OCCURRED ON         131361 non-null  object
 2   OCCURRED TO         93175 non-null   object
 3   UCR CRIME CATEGORY  131677 non-null  object
 4   100 BLOCK ADDR      131677 non-null  object
 5   ZIP                 131674 non-null  object
 6   PREMISE TYPE        130920 non-null  object
dtypes: object(7)
memory usage: 7.0+ MB


,INC NUMBER,OCCURRED ON,OCCURRED TO,UCR CRIME CATEGORY,100 BLOCK ADDR,ZIP,PREMISE TYPE
98422,"""201700000873158""",05/19/2017 22:00,05/20/2017 06:30,"""LARCENY-THEFT""","""64XX W PRESTON LN""","""85043""","""STREET / ROADWAY / ALLEY / SIDEWALK"""
37619,"""201600001079173""",06/10/2016 15:00,06/11/2016 14:30,"""BURGLARY""","""17XX W UNION HILLS DR""","""85027""","""CONDO / TOWNHOUSE"""
11289,"""201600000060020""",01/07/2016 11:30,01/07/2016 12:30,"""LARCENY-THEFT""","""24XX E BUCKEYE RD""","""85034""","""VEHICLE"""
636,"""201500002126332""",11/04/2015 19:37,11/04/2015 19:37,"""ROBBERY""","""11XX W BELL RD""","""85023""","""APARTMENT"""
55026,"""201680006476""",09/19/2016 00:00,09/20/2016 10:00,"""LARCENY-THEFT""","""XX E THOMAS RD""","""85012""","""PARKING LOT"""
110444,"""201700001291039""",07/24/2017 02:00,NaN,"""ROBBERY""","""132XX N 44TH ST""","""85032""","""PARK / PLAYGROUND"""
102356,"""201700001006695""",06/10/2017 01:35,NaN,"""AGGRAVATED ASSAULT""","""17XX W GLENDALE AVE""","""85015""","""STREET / ROADWAY / ALLEY / SIDEWALK"""
1009,"""201500002141634""",11/07/2015 03:36,11/07/2015 03:37,"""ROBBERY""","""122XX N CAVE CREEK RD""","""85022""","""CONVENIENCE MARKET / STORE"""
43309,"""201600001292312""",07/13/2016 16:10,NaN,"""LARCENY-THEFT""","""34XX E CHARLESTON AVE""","""85032""","""SINGLE FAMILY HOUSE"""
93753,"""201700000709011""",04/24/2017 06:00,04/24/2017 15:00,"""BURGLARY""","""7XX E PORTLAND ST""","""85006""","""OTHER"""


In [ ]:
df.shape

(131677, 7)

Data cleaning:  Change date string to datetime, check for na dates, and remove.

In [ ]:
df['OCCURRED ON'] = pd.to_datetime(df['OCCURRED ON'], errors='coerce') #convert to datetime objects
df['OCCURRED ON'].head()

,OCCURRED ON
0,2015-11-01
1,2015-11-01
2,2015-11-01
3,2015-11-01
4,2015-11-01


I removed some NA values from the "OCCURED ON" column

In [ ]:
df['OCCURRED ON'].isna().sum() #how many NaTs

np.int64(0)

In [ ]:
df = df[df['OCCURRED ON'].notna()] #get rid of them


More data cleaning to get ZIPs in correct format.

In [ ]:
df['ZIP'] = (
    df['ZIP']
    .astype(str) #everything is a string
    .str.replace('"', '', regex=False) #remove quotes
    .str.extract('(\d{5})')  # pull 5-digit ZIPs if extra text exists, if not Nan
)

<>:5: SyntaxWarning: invalid escape sequence '\d'
<>:5: SyntaxWarning: invalid escape sequence '\d'
/tmp/ipython-input-1698461913.py:5: SyntaxWarning: invalid escape sequence '\d'
  .str.extract('(\d{5})')  # pull 5-digit ZIPs if extra text exists, if not Nan


In [ ]:
df['ZIP'].isna().sum()

np.int64(3)

Since there were only 3 bad rows in zip ok to drop.

In [ ]:
df = df[df['ZIP'].astype(str).str.len() == 5]


In [ ]:
df['ZIP'].astype(str).str.len().value_counts()

,count
ZIP,
5,131358


Things to do per the assignment:
1. Grouping crimes by location type or by zip code (or groups of zip codes).
2. Or, on the contrary, looking at types of crimes and where they are most common.
3. Would be good to know which areas have the fastest growing and shrinking crime rates.
4. Might even be worth grouping crimes by violent and non-violent?

# Question 1. Group crimes by location type or by zip code (or groups of zip codes).


In [ ]:
df.columns.tolist()
df[['ZIP','PREMISE TYPE','UCR CRIME CATEGORY']].head()


,ZIP,PREMISE TYPE,UCR CRIME CATEGORY
0,85029,"""SINGLE FAMILY HOUSE""","""MOTOR VEHICLE THEFT"""
1,85006,"""SINGLE FAMILY HOUSE""","""RAPE"""
2,85014,"""APARTMENT""","""LARCENY-THEFT"""
3,85307,"""SINGLE FAMILY HOUSE""","""MOTOR VEHICLE THEFT"""
4,85043,"""SINGLE FAMILY HOUSE""","""MOTOR VEHICLE THEFT"""


Tidy premise type and UCR Crime category:

In [ ]:
df['PREMISE TYPE'] = df['PREMISE TYPE'].fillna('UNKNOWN').str.strip() #missing data assigned unknown, removes any extra spaces
df['UCR CRIME CATEGORY'] = df['UCR CRIME CATEGORY'].astype(str).str.strip() #convert to string, removes extra spaces


Where (premise type) are crimes happening most?


In [ ]:
crime_by_premise = (
    df.groupby('PREMISE TYPE')
      .size()
      .sort_values(ascending=False)
      .rename('count')
)

crime_by_premise.head(20)
df['PREMISE TYPE'].value_counts()

,count
PREMISE TYPE,
"""SINGLE FAMILY HOUSE""",25280
"""APARTMENT""",14936
"""PARKING LOT""",14121
"""STREET / ROADWAY / ALLEY / SIDEWALK""",11567
"""RETAIL BUSINESS""",10825
...,...
"""GAMBLING FACILITY/CASINO/RACE TRACK""",2
"""LAKE / WATERWAY / BEACH""",1
"""TRIBAL LANDS""",1


Crime counts group by zip code:

In [ ]:
crime_by_zip = (
    df.groupby('ZIP')
      .size()
      .sort_values(ascending=False)
      .rename('count')
)

crime_by_zip.head(20)
df['ZIP'].value_counts()

,count
ZIP,
85015,6923
85008,6308
85009,5490
85041,5471
85051,5364
...,...
85337,1
85338,1
85335,1


# Question 2.  Look at types of crimes and where they are most common.

In [ ]:
crime_counts = df['UCR CRIME CATEGORY'].value_counts()
print(crime_counts.to_string())

UCR CRIME CATEGORY
"LARCENY-THEFT"                            66041
"BURGLARY"                                 23603
"MOTOR VEHICLE THEFT"                      14367
"DRUG OFFENSE"                              9553
"AGGRAVATED ASSAULT"                        9273
"ROBBERY"                                   5624
"RAPE"                                      1976
"ARSON"                                      632
"MURDER AND NON-NEGLIGENT MANSLAUGHTER"      289


For each of these crime categories and premise types, where are crimes most likely to occur?  Edited to only show the top from each category.

In [ ]:
top_crimes = df['UCR CRIME CATEGORY'].value_counts().head(10).index
top_crimes


Index(['"LARCENY-THEFT"', '"BURGLARY"', '"MOTOR VEHICLE THEFT"',
       '"DRUG OFFENSE"', '"AGGRAVATED ASSAULT"', '"ROBBERY"', '"RAPE"',
       '"ARSON"', '"MURDER AND NON-NEGLIGENT MANSLAUGHTER"'],
      dtype='object', name='UCR CRIME CATEGORY')

In [ ]:
top_premises = df['PREMISE TYPE'].value_counts().head(10).index
top_premises

Index(['"SINGLE FAMILY HOUSE"', '"APARTMENT"', '"PARKING LOT"',
       '"STREET / ROADWAY / ALLEY / SIDEWALK"', '"RETAIL BUSINESS"',
       '"VEHICLE"', '"DRIVEWAY"', '"CONVENIENCE MARKET / STORE"',
       '"DEPARTMENT / DISCOUNT STORE"', '"GROCERY / SUPER MARKET"'],
      dtype='object', name='PREMISE TYPE')

In [ ]:
df_filter = df[
    df['UCR CRIME CATEGORY'].isin(top_crimes) &
    df['PREMISE TYPE'].isin(top_premises)
]  #filter out top

In [ ]:
ct_small = pd.crosstab(
    df_filter['UCR CRIME CATEGORY'],
    df_filter['PREMISE TYPE']
)
ct_small #creat crosstab table.  Looks messy.

PREMISE TYPE,"""APARTMENT""","""CONVENIENCE MARKET / STORE""","""DEPARTMENT / DISCOUNT STORE""","""DRIVEWAY""","""GROCERY / SUPER MARKET""","""PARKING LOT""","""RETAIL BUSINESS""","""SINGLE FAMILY HOUSE""","""STREET / ROADWAY / ALLEY / SIDEWALK""","""VEHICLE"""
UCR CRIME CATEGORY,,,,,,,,,,
"""AGGRAVATED ASSAULT""",2159,155,33,76,40,723,73,2026,2051,222
"""ARSON""",46,0,3,14,2,44,4,82,24,28
"""BURGLARY""",3890,321,84,22,41,77,656,12582,19,112
"""DRUG OFFENSE""",563,143,70,28,75,1205,323,565,2717,1432
"""LARCENY-THEFT""",5511,2900,3446,3491,3298,7114,9350,6824,2754,4304
"""MOTOR VEHICLE THEFT""",1861,105,33,932,49,4069,184,2301,2390,325
"""MURDER AND NON-NEGLIGENT MANSLAUGHTER""",44,3,0,3,0,38,4,64,54,8
"""RAPE""",521,4,2,0,0,51,4,607,117,124
"""ROBBERY""",341,627,122,39,104,800,227,229,1441,273


Gemini offered to code and run this for me:

In [ ]:
# Top 5 premises for each offense (by count)
counts = (
    df.groupby('UCR CRIME CATEGORY')['PREMISE TYPE']
      .value_counts()                            # counts per (crime, premise)
)

# Keep the top 5 premises within each offense
top5_per_offense = (
    counts.groupby(level=0)                      # level 0 = UCR CRIME CATEGORY
          .head(5)
          .reset_index(name='count')
)
top5_per_offense.head()


,UCR CRIME CATEGORY,PREMISE TYPE,count
0,"""AGGRAVATED ASSAULT""","""APARTMENT""",2159
1,"""AGGRAVATED ASSAULT""","""STREET / ROADWAY / ALLEY / SIDEWALK""",2051
2,"""AGGRAVATED ASSAULT""","""SINGLE FAMILY HOUSE""",2026
3,"""AGGRAVATED ASSAULT""","""PARKING LOT""",723
4,"""AGGRAVATED ASSAULT""","""PARK / PLAYGROUND""",233


In [ ]:
display(top5_per_offense)

,UCR CRIME CATEGORY,PREMISE TYPE,count
0,"""AGGRAVATED ASSAULT""","""APARTMENT""",2159
1,"""AGGRAVATED ASSAULT""","""STREET / ROADWAY / ALLEY / SIDEWALK""",2051
2,"""AGGRAVATED ASSAULT""","""SINGLE FAMILY HOUSE""",2026
3,"""AGGRAVATED ASSAULT""","""PARKING LOT""",723
4,"""AGGRAVATED ASSAULT""","""PARK / PLAYGROUND""",233
5,"""ARSON""",UNKNOWN,260
6,"""ARSON""","""SINGLE FAMILY HOUSE""",82
7,"""ARSON""","""APARTMENT""",46
8,"""ARSON""","""PARKING LOT""",44
9,"""ARSON""","""VEHICLE""",28


I had chatgpt run a summary of data, and I edited output:

The table above shows where different crime types most frequently occur across Phoenix.

Aggravated assaults are concentrated primarily in apartments, street/roadway areas, and single-family homes, suggesting that interpersonal violence occurs most often in residential environments and public spaces. Arson incidents are dominated by cases with unknown locations, followed by single-family homes and apartments, indicating that arson scenes are often either unclassified or involve residential structures.

Burglary targets single-family houses, which account for more than half of all burglary incidents, with smaller numbers in apartments and commercial buildings.

In contrast, drug offenses are strongly associated with streets, vehicles, and parking lots.

Larceny-theft, the most common crime category, is heavily concentrated in retail businesses and parking lots, consistent with patterns of shoplifting, vehicle break-ins, and thefts in commercial areas.

For motor vehicle theft, the dominant location is parking lots, followed by streets and homes with driveways, suggesting that both public and private parking areas present significant vulnerability.

Violent crimes such as murder and rape show a strong residential component, particularly single-family homes and apartments, but also appear in street or public locations, indicating both domestic and public-space risk patterns.

Finally, robberies most commonly occur in street/roadway areas, parking lots, and convenience markets, reinforcing their concentration in public and retail environments.

Overall, crime locations vary systematically by offense type:

Residential crimes dominate burglary, rape, and part of aggravated assault.

Public-space crimes dominate robbery, drug offenses, and many assaults.

Commercial areas show high concentrations of theft.

Parking lots emerge as a consistent hotspot across multiple crime categories.

These patterns can help guide targeted prevention strategies, including focused police patrols, improved lighting, and enhanced security measures in the most affected premise types.

# Question 3.  Which areas have the fastest growing and shrinking crime rates.

Period available:  Nov 2015 to Nov 2017

In [ ]:
df['month'] = df['OCCURRED ON'].dt.to_period('M')
df['month'].unique()

<PeriodArray>
['2015-11', '2015-12', '2016-01', '2016-02', '2016-03', '2016-04', '2016-05',
 '2016-06', '2016-07', '2016-08', '2016-09', '2016-10', '2016-11', '2016-12',
 '2017-01', '2017-02', '2017-03', '2017-04', '2017-05', '2017-06', '2017-07',
 '2017-08', '2017-09', '2017-10', '2017-11']
Length: 25, dtype: period[M]

In [ ]:
df['month'].value_counts().sort_index()

,count
month,
2015-11,4847
2015-12,5277
2016-01,5307
2016-02,4745
2016-03,5109
2016-04,5274
2016-05,5259
2016-06,5270
2016-07,5159


Crimes per zip per month:

In [ ]:
trend_zip = (
    df.groupby(['ZIP', 'month'])
      .size()
      .reset_index(name='count')
      .sort_values(['ZIP', 'month'])
)
trend_zip

,ZIP,month,count
0,85003,2015-11,69
1,85003,2015-12,65
2,85003,2016-01,61
3,85003,2016-02,56
4,85003,2016-03,41
...,...,...,...
1495,85395,2016-01,1
1496,85395,2016-02,1
1497,85396,2017-07,1
1498,85396,2017-08,2


Gemini offered to run code for me if I explained what I wanted.  This was my input:  "Identify the areas with the fastest growing and shrinking crime rates.  Calculate month to month percent changes in crime counts for each ZIP.  Average the change over 11/2015 to 11/2017. Sort by descending values."  ZIP codes with the highest average monthly percent increases represent the fastest-growing areas, while ZIPs with the most negative average changes represent the fastest-shrinking ZIPs.

In [ ]:
# Calculate month-to-month percentage change for each zip code
trend_zip['pct_change'] = trend_zip.groupby('ZIP')['count'].pct_change()

# Remove the first month for each zip code as it will have NaN for percent change
trend_zip_filtered = trend_zip.dropna(subset=['pct_change']).copy()

# Filter out zip codes with a low total number of crimes across the period
# (e.g., less than a certain threshold, let's use 50 as a starting point)
zip_counts = trend_zip_filtered.groupby('ZIP')['count'].sum()
zips_to_keep = zip_counts[zip_counts >= 50].index

trend_zip_filtered = trend_zip_filtered[trend_zip_filtered['ZIP'].isin(zips_to_keep)]

# Calculate the average monthly percentage change for each zip code
average_pct_change = trend_zip_filtered.groupby('ZIP')['pct_change'].mean().sort_values(ascending=False)

print("ZIP codes with the fastest growing crime rates (average monthly % change):")
display(average_pct_change.head())

print("\nZIP codes with the fastest shrinking crime rates (average monthly % change):")
display(average_pct_change.tail())

ZIP codes with the fastest growing crime rates (average monthly % change):


,pct_change
ZIP,
85251,0.342532
85087,0.301515
85045,0.259420
85083,0.209272
85331,0.184659



ZIP codes with the fastest shrinking crime rates (average monthly % change):


,pct_change
ZIP,
85035,-0.013389
85029,-0.013978
85023,-0.016266
85031,-0.019393
85042,-0.019762


The analysis shows that crime is rising most rapidly in ZIP codes 85251, 85087, 85045, 85083, and 85331, with average monthly increases ranging from +18% to +34%. ZIP codes 85035, 85029, 85023, 85031, and 85042 exhibit small but steady monthly decreases in crime (around −1–2%) over the same time period.

I asked chapgpt to tell me where these areas were on Phoenix, which it did, then it offered to draw a map showing me these areas.  I felt this was beyond the scope of this project, but would be an interesting analysis.

# Question 4.  Group crimes by violent and non-violent:

Clean up first, and define:

In [ ]:
df['UCR CRIME CATEGORY'] = (
    df['UCR CRIME CATEGORY']
      .str.strip()
      .str.replace('"', '', regex=False)
)

In [ ]:
violent = [
    "AGGRAVATED ASSAULT",
    "ROBBERY",
    "RAPE",
    "MURDER AND NON-NEGLIGENT MANSLAUGHTER"
]



In [ ]:
df['VIOLENCE'] = df['UCR CRIME CATEGORY'].isin(violent).map({
    True: 'Violent',
    False: 'Non-Violent'
})
df['UCR CRIME CATEGORY'].unique()

array(['MOTOR VEHICLE THEFT', 'RAPE', 'LARCENY-THEFT', 'DRUG OFFENSE',
       'BURGLARY', 'AGGRAVATED ASSAULT',
       'MURDER AND NON-NEGLIGENT MANSLAUGHTER', 'ROBBERY', 'ARSON'],
      dtype=object)

In [ ]:
df['VIOLENCE'].value_counts()

,count
VIOLENCE,
Non-Violent,114196
Violent,17162


Violent crimes represent approximately ~13.1% of all offenses, while non-violent crimes account for ~86.9% of incidents across Phoenix.  Math performed on calculator.

Save to .csv to use in Module 3 Essentials project.

In [ ]:
df.to_csv("crime_clean.csv", index=False)

In [ ]:
from google.colab import files
files.download("crime_clean.csv")


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>